diff --git a/src/game/paint/artboard.entity.js b/src/game/paint/artboard.entity.js
index e114a6a..5a8f3b3 100644
--- a/src/game/paint/artboard.entity.js
+++ b/src/game/paint/artboard.entity.js
@@ -1,62 +1,56 @@
 import { CANVAS_HEIGHT, CANVAS_WIDTH } from "$nissy-girl/screens/screen.consts.js";
 
+/** @import { Container } from "pixi.js" */
+/** @import { ContainerComponent } from "$game/shared/entity/world.entity.js" */
+
 import { COLOR_WHITE } from "$nissy-girl/screens/render.consts.js";
 
 import { GraphicsContext, Graphics } from "pixi.js";
 
-import { createWorld } from "$game/shared/entity/world.entity.js";
+import { createEntity } from "$game/shared/entity/entity.js";
 
-/**
- * @param {object} [options]
- * @param {number} [options.backgroundColor]
- * @param {number} options.width
- * @param {number} options.height
- * @returns {import("$game/shared/entity/world.entity.js").WorldEntity}
- */
-export const createArtboard = ({
+const createArtboardComponent = ({
+	pixels,
 	backgroundColor = COLOR_WHITE,
-	width : artboardWidth = CANVAS_WIDTH,
-	height : artboardHeight = CANVAS_HEIGHT,
-} = false) => {
-	const world = createWorld({
-		id : "artboard",
-	});
-
-	const pixels = new GraphicsContext();
-
-	const renderable = new Graphics(pixels);
-
-	world.getRenderable().addChild(renderable);
-
-	world.artboard = {
-		getContext() {
-			return pixels;
-		},
+	width = CANVAS_WIDTH,
+	height = CANVAS_HEIGHT,
+}) => {
+	return ({
 		clear() {
 			pixels
-				.rect(0, 0, artboardWidth, artboardHeight)
+				.clear()
+				.rect(0, 0, width, height)
 				.fill(backgroundColor);
+
+			return true;
+		},
+		getContext() {
+			return pixels;
 		},
-	};
+	});
+};
 
-	let cursor;
+/**
+ * @returns {import("$game/shared/entity/entity.js").Entity}
+ */
+export const createArtboard = () => {
+	const pixels = new GraphicsContext();
 
-	Object.defineProperty(world, "cursor", {
-		get() {
-			if(!cursor) {
-				for(const entity of world.getEntities()) {
-					if(entity.id === "cursor") {
-						cursor = entity;
-					}
-				}
-			}
+	const renderable = new Graphics(pixels);
 
-			return cursor;
+	const artboard = createEntity({
+		id : "artboard",
+		components : {
+			artboard : createArtboardComponent({ pixels }),
+			render : {
+				getRenderable() {
+					return renderable;
+				},
+			},
 		},
-		enumerable : true,
 	});
 
-	world.artboard.clear();
+	artboard.artboard.clear();
 
-	return world;
+	return artboard;
 };
diff --git a/src/game/paint/cursor.entity.js b/src/game/paint/cursor.entity.js
index d1fca8d..3a91b0e 100644
--- a/src/game/paint/cursor.entity.js
+++ b/src/game/paint/cursor.entity.js
@@ -1,107 +1,72 @@
-import { DPAD_DOWN, DPAD_LEFT, DPAD_RIGHT, DPAD_UP } from "$game/shared/input.consts.js";
-
-import { input } from "$nissy-girl/input.js";
-
-import { Sprite } from "pixi.js";
+import { Sprite, Assets } from "pixi.js";
 
 import moveUrl from "./assets/cursor-move.png";
 import stationaryUrl from "./assets/cursor-stationary.png";
+import { createEntity } from "$game/shared/entity/entity.js";
 
 import { createPencil } from "./tools/pencil.js";
-
-/** @import { WorldEntity } from "$src/game/shared/entity/world.entity.js" */
+import { createMovement } from "$game/shared/component/movement.js";
 
 const SPEED = 0.35;
 
-const DIRECTION = new Map([
-	[ DPAD_LEFT, -1 ],
-	[ DPAD_RIGHT, 1 ],
-	[ DPAD_UP, -1 ],
-	[ DPAD_DOWN, 1 ],
-]);
+/** @import { WorldEntity } from "$src/game/shared/entity/world.entity.js" */
 
 /**
- * @param world
+ * @param {object} options
+ * @param {WorldEntity} options.world
  * @returns {import("$game/shared/entity/scene.entity.js").Entity}
  */
-export const createCursor = (world) => {
-	let x = 50;
-	let y = 50;
-
+export const createCursor = ({
+	world,
+}) => {
 	const sprite = new Sprite();
+	const textures = {};
 
-	let textures;
+	const movement = createMovement({
+		x : 50,
+		y : 50,
+		speed : SPEED,
+	});
 
-	const moveDir = new Set();
+	const { artboard } = world.world.get("artboard");
 
-	const tool = createPencil(world);
+	const pencil = createPencil({ artboard, movement });
 
-	return {
+	return createEntity({
 		id : "cursor",
-		tool,
-		getRenderable() {
-			return sprite;
-		},
-		start() {},
-		stop() {},
-		hasUpdate() {
-			const dirCount = moveDir.size;
-
-			for(const dir of DIRECTION.keys()) {
-				if(input.state[dir]) {
-					moveDir.add(dir);
-				} else if(!input.state[dir]) {
-					moveDir.delete(dir);
-				}
-			}
-
-			return tool.hasUpdate()
-				|| moveDir.size !== dirCount
-				|| moveDir.size > 0;
-		},
-		update(dt) {
-			let didUpdate = false;
-
-			for(const dir of moveDir) {
-				if(dir === DPAD_LEFT || dir === DPAD_RIGHT) {
-					x += DIRECTION.get(dir) * dt * SPEED;
-
-					didUpdate = true;
-				}
-
-				if(dir === DPAD_DOWN || dir === DPAD_UP) {
-					y += DIRECTION.get(dir) * dt * SPEED;
-
-					didUpdate = true;
-				}
-			}
-
-			return tool.update() || didUpdate;
-		},
-		getPosition() {
-			return {
-				x : Math.round(x + 2.9),
-				y : Math.round(y + 2.9),
-			};
-		},
-		getTextureRequests() {
-			return { move : moveUrl, stationary : stationaryUrl };
-		},
-		setTextures(loadedTextures) {
-			textures = loadedTextures;
-
-			sprite.texture = textures.stationary;
-		},
-		render() {
-			sprite.texture = textures[moveDir.size > 0 ? "move" : "stationary"];
-
-			sprite.x = Math.round(x);
-			sprite.y = Math.round(y);
-
-			tool.render();
-		},
-		destroy() {
-			sprite.destroy();
+		components : {
+			movement,
+
+			tool : pencil,
+
+			render : {
+				async load() {
+					const move = await Assets.load(moveUrl);
+					const stationary = await Assets.load(stationaryUrl);
+
+					textures.move = move;
+					textures.stationary = stationary;
+
+					sprite.texture = stationary;
+
+					return true;
+				},
+				update() {
+					sprite.texture = textures[movement.isMoving() ? "move" : "stationary"];
+
+					sprite.x = Math.round(movement.getX());
+					sprite.y = Math.round(movement.getY());
+
+					// tool component implements its own render we delegate to in cursor render
+					pencil.render();
+				},
+				getRenderable() {
+					return sprite;
+				},
+				destroy() {
+					sprite.destroy();
+				},
+			},
 		},
-	};
+	});
 };
diff --git a/src/game/paint/paint.machine.js b/src/game/paint/paint.machine.js
index f0e0f45..ea9aaba 100644
--- a/src/game/paint/paint.machine.js
+++ b/src/game/paint/paint.machine.js
@@ -3,10 +3,10 @@ import { createMachine } from "xstate";
 import { stateLogger } from "$util/state-logger.actor.js";
 
 import { invokeScene } from "$game/shared/scene.actor.js";
-import { invokeInput } from "$game/shared/input.actor.js";
+import { invokeInput, invokeComponentInputListener } from "$game/shared/input.actor.js";
 import { gameloop } from "$game/shared/game-loop.machine.js";
 import { createCursor } from "./cursor.entity.js";
-import { sceneAction } from "$game/shared/scene-action.js";
+import { sceneAction, withScene } from "$game/shared/scene-action.js";
 
 import {
 	BUTTON_A,
@@ -37,50 +37,48 @@ export const paintMachine = createMachine({
 			invoke : [
 				invokeScene({
 					id : "drawing",
-					world : createArtboard,
 					entities : [
+						createArtboard,
 						createCursor,
 					],
+					componentOrder : [
+						"world",
+						"movement",
+						"tool",
+						"render",
+					],
 				}),
 				invokeInput,
+
+				invokeComponentInputListener(
+					withScene(
+						// this is wild...
+						(scene) => scene.world.world.get("cursor").movement,
+					),
+				),
 			],
 
 			on : {
 				[BUTTON_START] : {
-					actions : sceneAction(({ world }) => {
-						world.artboard.clear();
-						world.cursor.tool.stop();
-					}),
-				},
-			},
-
-			initial : "pen up",
+					actions : sceneAction((_, { world }) => {
+						const artboard = world.world.get("artboard");
+						const cursor = world.world.get("cursor");
 
-			states : {
-				"pen up" : {
-					on : {
-						[BUTTON_A] : {
-							actions : sceneAction(({ world }) => world.cursor.tool.start()),
-							target : "pen down",
-						},
-					},
+						artboard.artboard.clear();
+						cursor.tool.stop();
+					}),
 				},
 
-				"pen down" : {
-					on : {
-						[BUTTON_A] : [
-							{
-								guard : ({ event }) => event.state === RELEASED,
-								actions : sceneAction(({ world }) => world.cursor.tool.stop()),
-								target : "pen up",
-							},
-						],
-					},
+				[BUTTON_A] : {
+					actions : sceneAction(({ event }, { world }) => {
+						const cursor = world.world.get("cursor");
 
-					// hold processes from input directional repeat
-					always : {
-						actions : sceneAction(({ world }) => world.cursor.tool.start()),
-					},
+						if(event.state === RELEASED) {
+							cursor.tool.stop();
+						} else if(!cursor.tool.active) {
+							cursor.tool.begin();
+						}
+					}),
 				},
 			},
 		},
diff --git a/src/game/paint/tools/pencil.js b/src/game/paint/tools/pencil.js
index f1b6f45..5ad5818 100644
--- a/src/game/paint/tools/pencil.js
+++ b/src/game/paint/tools/pencil.js
@@ -2,29 +2,44 @@ import { COLOR_BLACK } from "$nissy-girl/screens/render.consts.js";
 
 import { coordsDiffer } from "$game/shared/component/position.js";
 
-export const createPencil = (world) => {
+export const createPencil = ({ artboard, movement }) => {
 	let isDrawing = false;
 	const pos = [];
 
 	return {
-		start() {
+		get active() {
+			return isDrawing;
+		},
+
+		begin() {
 			isDrawing = true;
 		},
+
+		// intentional scene lifecycle hook
 		stop() {
 			isDrawing = false;
 			pos.length = 0;
 		},
+
 		hasUpdate() {
-			return isDrawing || pos.length;
+			return isDrawing;
 		},
+
 		update() {
-			if(world.cursor && isDrawing) {
-				pos.push(world.cursor.getPosition());
+			if(!isDrawing) {
+				return false;
+			}
+
+			const nextPos = movement.getPosition();
+
+			if(pos.length === 0 || coordsDiffer(nextPos, pos.at(-1))) {
+				pos.push(nextPos);
 			}
 		},
+
 		render() {
 			if(isDrawing || pos.length) {
-				const pixels = world.artboard.getContext();
+				const pixels = artboard.getContext();
 
 				if(pos.length === 1) {
 					pixels
diff --git a/src/game/shared/component/movement.js b/src/game/shared/component/movement.js
new file mode 100644
index 0000000..9a65ca3
--- /dev/null
+++ b/src/game/shared/component/movement.js
@@ -0,0 +1,81 @@
+import { DPAD_DOWN, DPAD_LEFT, DPAD_RIGHT, DPAD_UP } from "$game/shared/input.consts.js";
+
+import { input } from "$nissy-girl/input.js";
+
+/** @import { WorldEntity } from "$src/game/shared/entity/world.entity.js" */
+
+const DIRECTION = new Map([
+	[ DPAD_LEFT, -1 ],
+	[ DPAD_RIGHT, 1 ],
+	[ DPAD_UP, -1 ],
+	[ DPAD_DOWN, 1 ],
+]);
+
+export const createMovement = ({
+	x : startX = 50,
+	y : startY = 50,
+	speed = 1,
+}) => {
+	// own position for the moment but likely pull out in the future
+	let x = startX;
+	let y = startY;
+
+	const moveDir = new Set();
+
+	const isMoving = () => moveDir.size > 0;
+
+	return ({
+		isMoving() {
+			return isMoving();
+		},
+
+		hasUpdate() {
+			return isMoving();
+		},
+
+		handleInput() {
+			const dirCount = moveDir.size;
+
+			for(const dir of DIRECTION.keys()) {
+				if(input.state[dir]) {
+					moveDir.add(dir);
+				} else if(!input.state[dir]) {
+					moveDir.delete(dir);
+				}
+			}
+
+			return moveDir.size !== dirCount
+				|| isMoving();
+		},
+
+		update(dt) {
+			if(!isMoving()) {
+				return false;
+			}
+
+			for(const dir of moveDir) {
+				if(dir === DPAD_LEFT || dir === DPAD_RIGHT) {
+					x += DIRECTION.get(dir) * dt * speed;
+				}
+
+				if(dir === DPAD_DOWN || dir === DPAD_UP) {
+					y += DIRECTION.get(dir) * dt * speed;
+				}
+			}
+
+			return true;
+		},
+
+		getPosition() {
+			return { x, y };
+		},
+
+		getX() {
+			return x;
+		},
+
+		getY() {
+			return y;
+		},
+	});
+};
diff --git a/src/game/shared/entity/entity.js b/src/game/shared/entity/entity.js
index 1607ccf..cfa623e 100644
--- a/src/game/shared/entity/entity.js
+++ b/src/game/shared/entity/entity.js
@@ -1,16 +1,45 @@
-import { Sprite, Container } from "pixi.js";
+/**
+ * @typedef {object} Component
+ * @property {() => void} start
+ * @property {() => void} stop
+ * @property {() => void} update
+ * @property {() => void} handleInput
+ */
 
 /**
  * @typedef {object} Entity
- * @property {() => void} start lifecycle
- * @property {() => void} stop lifecycle
- * @property {() => boolean} hasUpdate if should update
- * @property {() => void} update lifecycle
- * @property {() => void} render lifecycle
- * @property {() => void} destroy lifecycle
- * @property {() => Container} getRenderable get renderable for entity
- * @property {() => Record<string, string>} [getTextureRequests] optional: sprite key -> image url this entity needs
- * @property {(sprites: Record<string, Sprite>) => void} [setTextures] optional: receives loaded sprites, keyed the same as getSpriteRequests
+ * @property {boolean} hasComponent
+ * @property {Component} getComponent
  */
 
-// this will exist at some point probably, cursor kinda just does this for now
+/**
+ *
+ * @param {object} options
+ * @param {string} options.id
+ * @param {object} options.components
+ * @returns {Entity}
+ */
+export const createEntity = ({ id, components = {} }) => {
+	// TODO validate components to scene
+	const componentMap = new Map(Object.entries(components));
+
+	const entity = ({
+		id,
+		components,
+		getComponents() {
+			return componentMap;
+		},
+		hasComponent(id) {
+			return componentMap.has(id);
+		},
+		getComponent(id) {
+			return componentMap.get(id);
+		},
+	});
+
+	for(const [ id, component ] of componentMap) {
+		entity[id] = component;
+	}
+
+	return entity;
+};
diff --git a/src/game/shared/entity/scene.entity.js b/src/game/shared/entity/scene.entity.js
index f431180..b564530 100644
--- a/src/game/shared/entity/scene.entity.js
+++ b/src/game/shared/entity/scene.entity.js
@@ -1,6 +1,4 @@
-import { noopFalseFunction } from "$util/noop.js";
-
-import { Assets } from "pixi.js";
+import { createWorld } from "./world.entity.js";
 
 /** @import { Entity } from "./entity.js" */
 /** @import { WorldEntity } from "./world.entity.js" */
@@ -16,93 +14,95 @@ import { Assets } from "pixi.js";
  * @property {WorldEntity} world
  */
 
+const createUnrolledCallableByLifecycle = (components, entities, lifecycles) => {
+	const lifecycleMap = new Map(
+		lifecycles.map(
+			(lifecycle) => [ lifecycle, []],
+		),
+	);
+
+	// trying to only do this insane loop once
+	for(const component of components) {
+		for(const entity of entities) {
+			const entityComponents = entity.getComponents();
+
+			for(const lifecycle of lifecycles) {
+				if(entityComponents.get(component)?.[lifecycle]) {
+					lifecycleMap.get(lifecycle).push([ entity, entity[component][lifecycle] ]);
+				}
+			}
+		}
+	}
+
+	return lifecycleMap;
+};
+
 /**
  *
  * @param {object} options options obj
  * @param {string} options.id entity id
  * @param {() => WorldEntity} options.world
  * @param {(() => Entity)[]} options.entities child entities in canonical rendering canonical ordering
- * @param {() => void} options.start lifecycle
- * @param {() => void} options.stop lifecycle
+ * @param {string[]} options.componentOrder
  * @returns {SceneEntity} scene entity
  */
 export const createScene = ({
 	id,
-	world : worldFactory,
+	world : worldFactory = createWorld,
 	entities : entityFactories,
-	start = noopFalseFunction,
-	stop = noopFalseFunction,
+	componentOrder,
 }) => {
-	let isRunning = false;
-
 	const world = worldFactory();
 	const entities = [ world ];
 
+	let isAlive = true;
+
 	for(const createEntity of entityFactories) {
-		const entity = createEntity(world);
+		const entity = createEntity({ world });
 
-		world.registerEntity(entity);
+		world.world.add(entity);
 
 		entities.push(entity);
 	}
 
+	const lifecycleMap = createUnrolledCallableByLifecycle(
+		componentOrder,
+		entities,
+		[
+			"load",
+			"destroy",
+			"stop",
+			"hasUpdate",
+			"update",
+		],
+	);
+
+	// set calls upfront so iteration has as few lookups as possible
+	const load = lifecycleMap.get("load").map(([ , loadFunc ]) => loadFunc);
+	const hasUpdate = lifecycleMap.get("hasUpdate").map(([ , hasUpdateFunc ]) => hasUpdateFunc);
+	const update = lifecycleMap.get("update");
+	const stop = lifecycleMap.get("stop").map(([ , stopFunc ]) => stopFunc);
+	const destroy = lifecycleMap.get("destroy").map(([ , destroyFunc ]) => destroyFunc);
+	// We clearly aren't really ECS but we're going to try and model that direction keep room to grow!
+
 	const scene = {
 		id,
-
 		world,
 
-		start() {
-			if(isRunning) {
-				throw new Error("Cannot start in progress scene", this);
-			}
-
-			isRunning = true;
-
-			for(const entity of entities) {
-				entity.start();
-			}
-
-			start(this);
+		// scene-action caches the scene while it's alive, expose lifecycle so it knows to get rid of us
+		get isAlive() {
+			return isAlive;
 		},
 
 		async load() {
 			return Promise.all(
-				entities
-					.filter((entity) => entity.getTextureRequests)
-					.map(async (entity) => {
-						const requests = entity.getTextureRequests();
-						const textures = {};
-
-						await Promise.all(
-							Object.entries(requests).map(async ([ key, url ]) => {
-								textures[key] = await Assets.load(url);
-							}),
-						);
-
-						entity.setTextures(textures);
-					}),
+				load.map((loadFunc) => loadFunc()),
 			);
 		},
 
-		destroy() {
-			for(const entity of entities) {
-				entity.destroy();
-			}
-		},
-
-		stop() {
-			isRunning = false;
-
-			for(const entity of entities) {
-				entity.stop();
-			}
-
-			stop(this);
-		},
-
 		hasUpdate() {
-			for(const entity of entities) {
-				if(entity.hasUpdate()) {
+			for(const hasUpdateFunc of hasUpdate) {
+				if(hasUpdateFunc()) {
 					return true;
 				}
 			}
@@ -111,17 +111,24 @@ export const createScene = ({
 		},
 
 		update(dt) {
-			for(const entity of entities) {
-				if(entity.hasUpdate()) {
-					entity.update(dt);
-				}
+			for(const [ entity, updateFunc ] of update) {
+				updateFunc(dt, entity);
+			}
+		},
+
+		// stop is maybe not semantically making sense so far
+		stop() {
+			for(const stopFunc of stop) {
+				stopFunc();
 			}
 		},
 
-		render() {
-			for(const entity of entities) {
-				entity.render();
+		destroy() {
+			for(const destroyFunc of destroy) {
+				destroyFunc();
 			}
+
+			isAlive = false;
 		},
 	};
 
diff --git a/src/game/shared/entity/world.entity.js b/src/game/shared/entity/world.entity.js
index 46e85b3..22f1ee6 100644
--- a/src/game/shared/entity/world.entity.js
+++ b/src/game/shared/entity/world.entity.js
@@ -1,67 +1,77 @@
 import { Container } from "pixi.js";
 
+import { createEntity } from "./entity.js";
+
 /** @import { Entity } from "./entity.js" */
 
 /**
- * @typedef {object} WorldEntity
- * @property {() => void} start lifecycle
- * @property {() => void} destroy lifecycle
- * @property {() => boolean} hasUpdate if should update
- * @property {() => void} update lifecycle
- * @property {() => void} stop pause
- * @property {() => Container} getRenderable get renderable for entity
- * @property {(Entity) => void} registerEntity register entity to world
+ * @typedef {object} Component
+ * @property {() => void} [update]
+ * @property {() => void} [start]
+ * @property {() => void} [stop]
+ * @property {() => void} [destroy]
  */
 
 /**
- * @param {object} [options]
- * @param {string} options.id
- * @param {number} options.width
- * @param {number} options.height
- * @returns {WorldEntity}
+ * @typedef {object} ContainerComponent
+ * @property {(Entity) => void} add
+ * \@property {() => void} remove
+ * @property {() => Container} getRenderable
+ * @property {() => void} destroy
  */
-export const createWorld = ({
-	id,
-	// todo
-	width = 1,
-	height = 1,
-} = false) => {
-	const surface = new Container();
 
+/**
+ * @returns {ContainerComponent}
+ */
+const createContainerComponent = () => {
 	const entities = [];
+	const entityMap = new Map();
+	const surface = new Container();
 
-	return {
-		id,
-
-		start() {},
-
-		registerEntity(entity) {
+	return ({
+		add(entity) {
 			entities.push(entity);
+			entityMap.set(entity.id, entity);
 
-			surface.addChild(entity.getRenderable());
-		},
+			if(entity?.render?.getRenderable) {
+				surface.addChild(entity.render.getRenderable());
+			}
 
-		// this may be antipattern
-		getEntities() {
-			return entities;
+			// components get a special reference to their world component!
+			entity.world = this;
 		},
 
-		hasUpdate() {
-			return false;
+		get(entityId) {
+			return entityMap.get(entityId);
 		},
 
-		update() {},
-
-		render() {},
-
+		// ??? no removal for the moment
 		getRenderable() {
 			return surface;
 		},
-
-		stop() {},
-
 		destroy() {
 			surface.destroy();
 		},
-	};
+	});
+};
+
+/**
+ * @param {object} [options]
+ * @param {string} options.id
+ * @param {object} options.components
+ * @returns {Entity}
+ */
+export const createWorld = ({
+	id = "world",
+	components = {},
+} = false) => {
+	const worldEntity = createEntity({
+		id,
+		components : {
+			world : createContainerComponent(),
+			...components,
+		},
+	});
+
+	return worldEntity;
 };
diff --git a/src/game/shared/game-loop.machine.js b/src/game/shared/game-loop.machine.js
index 09330a6..2737fea 100644
--- a/src/game/shared/game-loop.machine.js
+++ b/src/game/shared/game-loop.machine.js
@@ -28,12 +28,9 @@ export const gameloop = {
 				}
 
 				if(scene) {
-					if(scene.hasUpdate()) {
-						scene.update(dt);
-					}
+					scene.update(dt);
 
-					scene.render();
-					screen.render(scene.world.getRenderable());
+					screen.render(scene.world.world.getRenderable());
 				}
 
 				const endFrameState = scene?.hasUpdate?.() || hasInput;
@@ -150,7 +147,7 @@ export const gameloop = {
 								actions : [
 									({ context }) => {
 										context.loop.stop();
-										context.scene.stop();
+										context.scene.destroy();
 										screen.clear();
 									},
 									assign({ scene : false }),
diff --git a/src/game/shared/input.actor.js b/src/game/shared/input.actor.js
index fd50d03..05f2535 100644
--- a/src/game/shared/input.actor.js
+++ b/src/game/shared/input.actor.js
@@ -1,5 +1,6 @@
 import { fromCallback } from "xstate";
 import { input } from "$nissy-girl/input.js";
+import { isActorAlive } from "$util/is-actor-alive.js";
 
 import {
 	DPAD_DOWN,
@@ -22,17 +23,18 @@ const REPEATING_INPUTS = new Set([
 ]);
 
 export const invokeInput = {
+	systemId : "input",
 	id : "input",
-	src : fromCallback(({ sendBack, system }) => {
+	src : fromCallback(({ sendBack, system, receive }) => {
 		const gameloop = system.get("gameloop");
 
+		const componentTargets = new Set();
 		const isRepeating = new Set();
 
 		let repeat = 0;
 
 		gameloop.send({
 			type : "REGISTER_INPUT",
-
 			input : (dt) => {
 				repeat += dt;
 
@@ -52,6 +54,15 @@ export const invokeInput = {
 			// proxy triggered events back
 			sendBack(event);
 
+			for(const component of componentTargets) {
+				try{
+					component.handleInput();
+				} catch (e) {
+					/* eslint-disable-next-line */
+					console.warn("Attempted to send input to component with no handleInput...", e);
+				}
+			}
+
 			if(REPEATING_INPUTS.has(event.type)) {
 				if(event.state === RELEASED) {
 					isRepeating.delete(event.type);
@@ -65,6 +76,20 @@ export const invokeInput = {
 			}
 		});
 
+		receive((event) => {
+			if(event.type === "SEND_COMPONENT_INPUT") {
+				componentTargets.add(event.component);
+
+				return;
+			}
+
+			if(event.type === "REMOVE_COMPONENT_INPUT") {
+				componentTargets.add(event.component);
+
+				return;
+			}
+		});
+
 		return () => {
 			// make sure everyone knows input has ended
 			for(const repeating of isRepeating) {
@@ -73,9 +98,34 @@ export const invokeInput = {
 
 			removeInput();
 
-			if(gameloop.getSnapshot().status === "active") {
+			if(isActorAlive(gameloop)) {
 				gameloop.send({ type : "REMOVE_INPUT" });
 			}
 		};
 	}),
 };
+
+let counter = 0;
+
+export const invokeComponentInputListener = (func) => ({
+	id : `component-input-listener-${counter}`,
+	src : fromCallback((all) => {
+		const { system } = all;
+
+		const component = func(all);
+
+		if(!component) {
+			throw new Error("[invokeComponentInputListener] sent invalid component");
+		}
+
+		const input = system.get("input");
+
+		input.send({ type : "SEND_COMPONENT_INPUT", component });
+
+		return () => {
+			if(isActorAlive(input)) {
+				input.send({ type : "REMOVE_COMPONENT_INPUT", component });
+			}
+		};
+	}),
+});
diff --git a/src/game/shared/scene-action.js b/src/game/shared/scene-action.js
index 88e635c..85022aa 100644
--- a/src/game/shared/scene-action.js
+++ b/src/game/shared/scene-action.js
@@ -1,5 +1,31 @@
-export const sceneAction = (func) => ({ system }) => {
-	const scene = system.get("scene");
+// scene entity
+let activeSceneRef = false;
 
-	func(scene.getSnapshot().context);
+const getAndCacheScene = ({ system }) => {
+	if(!activeSceneRef || !activeSceneRef?.isAlive) {
+		const scene = system.get("scene");
+
+		activeSceneRef = scene.getSnapshot().context;
+	}
+
+	return activeSceneRef;
+};
+
+export const withScene = (func) => (all) => {
+	return func(getAndCacheScene(all));
+};
+
+/**
+ * sceneAction forces a frame unless returns `false
+ * @param {() => void} func func receives xstate and scene
+ * @returns {() => void}
+ */
+export const sceneAction = (func) => (all) => {
+	const result = func(all, getAndCacheScene(all));
+
+	if(result !== false) {
+		const gameloop = all.system.get("gameloop");
+
+		gameloop.send({ type : "START" });
+	}
 };
diff --git a/src/game/shared/scene.actor.js b/src/game/shared/scene.actor.js
index 5f52593..db1f7b1 100644
--- a/src/game/shared/scene.actor.js
+++ b/src/game/shared/scene.actor.js
@@ -1,6 +1,6 @@
 import { fromObservable } from "xstate";
 
-import { noopFalseFunction } from "$util/noop.js";
+import { isActorAlive } from "$util/is-actor-alive.js";
 
 import { createScene } from "./entity/scene.entity.js";
 
@@ -8,8 +8,7 @@ export const invokeScene = ({
 	id,
 	world,
 	entities,
-	start = noopFalseFunction,
-	stop = noopFalseFunction,
+	componentOrder,
 }) => ({
 	systemId : "scene",
 	id,
@@ -23,31 +22,28 @@ export const invokeScene = ({
 				id,
 				world,
 				entities,
-				start,
-				stop,
+				componentOrder,
 			});
 
+			observer.next(scene);
+
 			scene.load().then(() => {
 				if(cancelled) {
 					return;
 				}
 
-				scene.start();
-
 				gameloop.send({ type : "REGISTER_SCENE", scene });
 				gameloop.send({ type : "START" });
-
-				observer.next(scene);
 			});
 
 			return {
 				unsubscribe() {
 					cancelled = true;
 
-					if(gameloop.getSnapshot().status === "active") {
+					if(isActorAlive(gameloop)) {
 						gameloop.send({ type : "REMOVE_SCENE" });
 					} else {
-						scene.stop();
+						scene.destroy();
 					}
 				},
 			};
diff --git a/src/nissy-girl/input.js b/src/nissy-girl/input.js
index 002a816..489c5a7 100644
--- a/src/nissy-girl/input.js
+++ b/src/nissy-girl/input.js
@@ -46,8 +46,8 @@ const startKeyListeners = (fire) => {
 		}
 	};
 
-	subs.add("keydown", domListenerSub(window, "keydown", handleKeyPress(TRIGGERED, fire)));
-	subs.add("keyup", domListenerSub(window, "keyup", handleKeyPress(RELEASED, fire)));
+	subs.add("keydown", domListenerSub(window, "keydown", handleKeyPress(TRIGGERED)));
+	subs.add("keyup", domListenerSub(window, "keyup", handleKeyPress(RELEASED)));
 
 	return () => subs.removeAll();
 };
diff --git a/src/util/is-actor-alive.js b/src/util/is-actor-alive.js
new file mode 100644
index 0000000..1762c00
--- /dev/null
+++ b/src/util/is-actor-alive.js
@@ -0,0 +1,2 @@
+export const isActorAlive = (actor) =>
+	actor?.getSnapshot?.()?.status === "active";
